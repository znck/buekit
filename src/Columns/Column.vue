<script>
import PropTypes, { normalize } from '@znck/prop-types'
import { styleResolver, style } from '../utils'
import { devices, additionalDevices, columns as sizes } from '../shared'
import { createTag } from '../mixins'

const sizeType = (name, extras = []) => PropTypes.oneOfType(
  PropTypes.number.validate(v => 1 <= v && v <= 12),
  PropTypes.string.validate(v => (x => 1 <= x && x <= 12)(Number(x))),
  PropTypes.oneOf(sizes)
).modifiers(name, [].concat(devices, extras))

let normalizer
const props = (props, _) => {
  normalizer = normalizer || normalize(props, (name, value) => {
      switch (name) {
        case 'size': return [_(`is-${value}`)]
        case 'offset': return [_(`is-${name}-${value}`)]
        case 'narrow': return value && [_(`is-${name}`)]
        default: return value
      }
    }, (name, suffix, value) => {
      switch (name) {
        case 'size': return _(`is-${value}-${suffix}`)
        case 'narrow': return value && _(`is-${name}-${suffix}`)
        case 'offset': return _(`is-${name}-${value}-${suffix}`)
      }
    }
  )

  return normalizer(props)
}

export default {
  name: 'Column',
  functional: true,
  mixins: [createTag()],
  props: {
    /**
     * Take only the space it needs.
     * @name narrow
     * @version 0.0.0
     * @since Version 0.0.0
     * @modifiers mobile tablet desktop
     */
    ...PropTypes.bool.value(false).modifiers('narrow', devices),
    /**
     * Horizontal space before a column element.
     * @name offset
     * @version 0.0.0
     * @since Version 0.0.0
     * @modifiers mobile tablet desktop
     */
    ...sizeType('offset'),
    /**
     * Width of column element in 12 point grid system.
     * @name size
     * @version 0.0.0
     * @since Version 0.0.0
     * @modifiers mobile tablet desktop
     */
    ...sizeType('size', additionalDevices)
  },

  render (h, ctx) {
    const _ = styleResolver(ctx.$style)
    const { tag, ...rest } = props(ctx.props, _)

    return style(h(tag, ctx.data, ctx.children), _('column'), Object.values(rest))
  }
}
</script>

<style lang="scss" src="./columns.scss" module></style>
