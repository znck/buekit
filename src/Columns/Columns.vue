<script>
import PropType, { normalize } from '@znck/prop-types'
import { isBoolean } from 'lodash'

import { styleResolver, style } from '../utils'
import { createTag } from '../mixins'

let normalizer
const props = (props, _) => {
  normalizer = normalizer || normalize(props, (name, value) => {
    switch (name) {
      case 'gap': return value === false ?
        [_('is-gapless')] :
        (value !== true ? [_('is-variable'), _(`is-${value}`)] : [])
      case 'tag': return value
      default: return value && _(`is-${name}`)
    }
  })

  return normalizer(props)
}

export default {
  name: 'Columns',
  functional: true,
  mixins: [createTag()],
  props: {
    /**
     * Put column elements in center horizontally.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    centered: PropType.bool.value(false),
    /**
     * Put column elements in center vertically.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    vcentered: PropType.bool.value(false),
    /**
     * Activate columns from mobile up.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    mobile: PropType.bool.value(false),
    /**
     * Activate columns for desktop only.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    desktop: PropType.bool.value(false),
    /**
     * The space on each side of a column.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    gap: PropType.oneOfType(
      PropType.bool,
      PropType.number.validate(v => 0 <= v && v <= 8),
      PropType.string.validate(v => /^[0-8]$/.test(v))
    ).value(true),
    /**
     * Wrap column elements that would fit in a single row.
     * @version 0.0.0
     * @since Version 0.0.0
     */
    multiline: PropType.bool.value(false),
  },

  render (h, ctx) {
    const _ = styleResolver(ctx.$style)
    const { tag, ...rest } = props(ctx.props, _)

    return style(h(tag, ctx.data, ctx.children), _('columns'), Object.values(rest))
  }
}
</script>

<style lang="scss" src="./columns.scss" module></style>
