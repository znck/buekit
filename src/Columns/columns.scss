@import '~bulma/sass/utilities/_all.sass';
@import '../_base.scss';
@import '~bulma/sass/grid/columns.sass';
