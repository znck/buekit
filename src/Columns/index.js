import Columns from './Columns.vue'
import Column from './Column.vue'

export {
  Column,
  Columns
}

export const Row = Columns // Export as row for better api.
