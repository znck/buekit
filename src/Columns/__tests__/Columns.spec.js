import {shallow} from 'vue-test-utils'
import {createRenderer} from 'vue-server-renderer'

import Columns from '../Columns.vue'

const renderer = createRenderer()

function snapshots(options) {
	options.forEach(
    option => it('matches snapshot: ' + JSON.stringify(option), () => {
      const wrapper = shallow(Columns, {
        context: {props: option},
        slots: {default: '<p>content</p>'}
      })

      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) {
          throw new Error(err)
        }
        expect(str).toMatchSnapshot()
      })
    })
  )
}

describe('Columns.vue', () => {
	snapshots([
    {},
    {centered: true},
    {vcentered: true},
    {gap: false},
    {gap: 8},
    {mobile: true},
    {tablet: true},
    {desktop: true},
    {multiline: true}
	])
})
