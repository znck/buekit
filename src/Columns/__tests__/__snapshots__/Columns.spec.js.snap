// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns.vue matches snapshot: {"centered":true} 1`] = `<div class="columns is-centered"></div>`;

exports[`Columns.vue matches snapshot: {"desktop":true} 1`] = `<div class="columns is-desktop"></div>`;

exports[`Columns.vue matches snapshot: {"gap":8} 1`] = `<div class="columns is-variable is-8"></div>`;

exports[`Columns.vue matches snapshot: {"gap":false} 1`] = `<div class="columns is-gapless"></div>`;

exports[`Columns.vue matches snapshot: {"mobile":true} 1`] = `<div class="columns is-mobile"></div>`;

exports[`Columns.vue matches snapshot: {"multiline":true} 1`] = `<div class="columns is-multiline"></div>`;

exports[`Columns.vue matches snapshot: {"tablet":true} 1`] = `<div class="columns"></div>`;

exports[`Columns.vue matches snapshot: {"vcentered":true} 1`] = `<div class="columns is-vcentered"></div>`;

exports[`Columns.vue matches snapshot: {} 1`] = `<div class="columns"></div>`;
